### PR TITLE
fix(controller): handle `nil` `processedTmpl` in DAGs

### LINKED ISSUE #13547: controller panic: `nil pointer dereference` on DAG `processedTmpl`
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I have tested with the `:latest` image tag (i.e. `quay.io/argoproj/workflow-controller:latest`) and can confirm the issue still exists on `:latest`. If not, I have explained why, **in detail**, in my description below.
- [X] I have searched existing issues and could not find a match for this bug
- [X] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened? What did you expect to happen?

`kubectl get workflow`
```
NAME   STATUS   AGE   MESSAGE
a      Error    14m   runtime error: invalid memory address or nil pointer dereference
```

`kubectl get workflow a -o yaml`
```yaml
# ...
    outputs: {}
status:
  artifactGCStatus:
    notSpecified: true
  artifactRepositoryRef:
    artifactRepository: {}
    default: true
  conditions:
  - status: "False"
    type: PodRunning
  - status: "True"
    type: Completed
  finishedAt: "2024-09-02T08:21:01Z"
  message: 'runtime error: invalid memory address or nil pointer dereference'
  nodes:
# ...
```

### Version(s)

latest(sha256:f61c2eaf06294f38dc66f7519ae2c3af916ce7dcde8322999789bd39c29e983c)

### Paste a minimal workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: a
spec:
  templates:
    - name: echo2
      container:
        name: ''
        image: alpine:3.7
        command:
          - echo
          - abc
    - name: echo1
      inputs:
        parameters:
          - name: abc
            valueFrom:
              configMapKeyRef:
                name: abc
                key: abc
      container:
        name: ''
        image: alpine:3.7
        command:
          - echo
          - abc
    - name: diamond
      dag:
        tasks:
          - name: A
            template: echo1
            when: 1 == 0      
          - name: B
            template: echo2         
            dependencies:
              - A
  entrypoint: diamond
```

### Logs from the workflow controller

```text
time="2024-09-02T08:21:01.019Z" level=info msg="node changed" namespace=default new.message=PodInitializing new.phase=Pending new.progress=0/1 nodeID=a-814491500 old.message= old.phase=Pending old.progress=0/1 workflow=a
time="2024-09-02T08:21:01.019Z" level=error msg="Mark error node" error="unable to retrieve inputs.parameters.abc from ConfigMap: ConfigMap 'abc' does not exist. Please make sure it has the label workflows.argoproj.io/configmap-type: Parameter to be detectable by the controller" namespace=default nodeName=a.A workflow=a
time="2024-09-02T08:21:01.019Z" level=error msg="node is already fulfilled" fromPhase=Skipped namespace=default nodeName=a.A toPhase=Error workflow=a
time="2024-09-02T08:21:01.019Z" level=info msg="node a-864824357 phase Skipped -> Error" namespace=default workflow=a
time="2024-09-02T08:21:01.019Z" level=info msg="node a-864824357 message: unable to retrieve inputs.parameters.abc from ConfigMap: ConfigMap 'abc' does not exist. Please make sure it has the label workflows.argoproj.io/configmap-type: Parameter to be detectable by the controller" namespace=default workflow=a
time="2024-09-02T08:21:01.020Z" level=error msg="Recovered from panic" namespace=default r="runtime error: invalid memory address or nil pointer dereference" stack="goroutine 268 [running]:\nruntime/debug.Stack()\n\t/usr/local/go/src/runtime/debug/stack.go:24 +0x64\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate.func2()\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:196 +0x88\npanic({0x1dc2200?, 0x390cab0?})\n\t/usr/local/go/src/runtime/panic.go:770 +0x124\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0x40006f40c0, {0x2566fe0, 0x39ce820}, 0x400031e770, {0x40008ba421, 0x1})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:453 +0xeb8\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAGTask(0x40006f40c0, {0x2566fe0, 0x39ce820}, 0x400031e770, {0x3897ef0, 0x1})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:517 +0x2600\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeDAG(0x40006f40c0, {0x2566fe0, 0x39ce820}, {0x3897fe8, 0x1}, 0x400087e2c0, {0x4000a8e370, 0x7}, 0x4000cefd48, {0x256d550, ...}, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/dag.go:269 +0x31c\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).executeTemplate(0x40006f40c0, {0x2566fe0, 0x39ce820}, {0x3897fe8, 0x1}, {0x256d550, 0x4000972000}, 0x400087e240, {{0x0, 0x0, ...}, ...}, ...)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:2207 +0x2594\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*wfOperationCtx).operate(0x40006f40c0, {0x2566fe0, 0x39ce820})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/operator.go:372 +0x1744\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).processNextItem(0x4000471608, {0x2566fe0, 0x39ce820})\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:888 +0x57c\ngithub.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).runWorker(0x4000471608)\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:803 +0x88\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.15/pkg/util/wait/wait.go:157 +0x40\nk8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x40009614e0, {0x253b0e0, 0x40008d6150}, 0x1, 0x4000563b60)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.15/pkg/util/wait/wait.go:158 +0x90\nk8s.io/apimachinery/pkg/util/wait.JitterUntil(0x40009614e0, 0x3b9aca00, 0x0, 0x1, 0x4000563b60)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.15/pkg/util/wait/wait.go:135 +0x80\nk8s.io/apimachinery/pkg/util/wait.Until(...)\n\t/go/pkg/mod/k8s.io/apimachinery@v0.26.15/pkg/util/wait/wait.go:92\ncreated by github.com/argoproj/argo-workflows/v3/workflow/controller.(*WorkflowController).Run in goroutine 56\n\t/go/src/github.com/argoproj/argo-workflows/workflow/controller/controller.go:367 +0x128c\n" workflow=a
time="2024-09-02T08:21:01.020Z" level=info msg="Updated phase Running -> Error" namespace=default workflow=a
time="2024-09-02T08:21:01.020Z" level=info msg="Updated message  -> runtime error: invalid memory address or nil pointer dereference" namespace=default workflow=a
time="2024-09-02T08:21:01.020Z" level=info msg="Marking workflow completed" namespace=default workflow=a
time="2024-09-02T08:21:01.020Z" level=info msg="Workflow to be dehydrated" Workflow Size=2359
```

### Logs from in your workflow's wait container

```text
None
```

### PR DESCRIPTION
**Pull request by chengjoey**
Fixes #13547

### Motivation
It seems more reasonable to judge `processedTmpl != nil`

### CODE REVIEW

File: workflow/controller/dag.go [hunk 1]
```diff
@@ -449,7 +449,7 @@ func (woc *wfOperationCtx) executeDAGTask(ctx context.Context, dagCtx *dagContex
 		}
 
 		// Release acquired lock completed task.
-		if tmpl != nil {
```
Comment (Joibel) [inline workflow/controller/dag.go:452]: I agree, it looks like this was overlooked in #9405.

Did you consider whether there should actually be a guard higher up in the code against continuing this far if `tmpl == nil`? It seems odd that we don't `return` if line 446 returns error, in symmetry with line 437.

Comment (chengjoey) [reply workflow/controller/dag.go:452]: Yes, I think we should get the error at line 434, and if error is nil, then must`tmpl != nil`

```
_, tmpl, _, tmplErr := dagCtx.tmplCtx.ResolveTemplate(task)
if tmplErr != nil {
	woc.markNodeError(node.Name, tmplErr)
	return
}
```

File: workflow/controller/dag.go [hunk 2]
```diff
@@ -431,7 +431,11 @@ func (woc *wfOperationCtx) executeDAGTask(ctx context.Context, dagCtx *dagContex
 
 	if node != nil && node.Fulfilled() {
 		// Collect the completed task metrics
-		_, tmpl, _, _ := dagCtx.tmplCtx.ResolveTemplate(task)
+		_, tmpl, _, tmplErr := dagCtx.tmplCtx.ResolveTemplate(task)
+		if tmplErr != nil {
+			woc.markNodeError(node.Name, tmplErr)
+			return
+		}
```
[the remark this answers is not in the document]
Comment (Joibel) [reply workflow/controller/dag.go:438]: Yes, that looks good.

**Review by Joibel — CHANGES_REQUESTED** [review]
As you have a reproducible test case, could you add it as a regression test please?

**Comment by chengjoey** [comment]
> As you have a reproducible test case, could you add it as a regression test please?

@Joibel , I accidentally reproduced it in #13547, but I can't reproduce it 100% now. It seems to be a one-time thing, so I'm not sure if I can add e2e tests

**Review by terrytangyuan — APPROVED** [review]
I think this is a great fix for edge cases even though it cannot be reproduced easily. WDYT? @Joibel

**Review by Joibel — APPROVED** [review]

**Comment by Joibel** [comment]
@chengjoey Thanks for this
